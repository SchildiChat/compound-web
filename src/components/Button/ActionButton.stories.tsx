/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { Meta } from "@storybook/react";
import { fn } from "@storybook/test";

import * as icons from "@vector-im/compound-design-tokens/assets/web/icons";

import { Button as ButtonComponent } from "./Button";
import { Tooltip } from "../Tooltip/Tooltip";
import { TooltipProvider } from "../Tooltip/TooltipProvider";

const Template: React.FC<
  { label: string } & React.ComponentProps<typeof ButtonComponent>
> = ({ label, ...args }) => (
  <TooltipProvider>
    <Tooltip label={label}>
      <ButtonComponent iconOnly {...args} />
    </Tooltip>
  </TooltipProvider>
);

export default {
  title: "Button/Actions",
  component: Template,
  tags: ["autodocs"],
  argTypes: {
    label: {
      type: "string",
    },
    size: {
      options: ["sm", "lg"],
      control: { type: "inline-radio" },
    },
    kind: {
      options: ["primary", "secondary", "tertiary"],
      control: { type: "inline-radio" },
    },
    destructive: { type: "boolean" },
    disabled: {
      type: "boolean",
    },
    as: {
      options: ["a", "button"],
      control: { type: "inline-radio" },
    },
    Icon: {
      options: Object.keys(icons),
      mapping: icons,
      control: { type: "select" },
    },
    onClick: { action: "onClick" },
  },
  args: {
    label: "Action",
    size: "lg",
    as: "button",
    destructive: false,
    disabled: false,
    onClick: fn(),
  },
  controls: {
    exclude: ["children"],
  },
} as Meta<typeof Template>;

export const Settings = {
  args: {
    label: "Settings",
    kind: "secondary",
    Icon: icons.SettingsSolidIcon,
  },
};

export const ShareLink = {
  args: {
    label: "Copy link",
    kind: "secondary",
    Icon: icons.LinkIcon,
  },
};

export const VideoOn = {
  args: {
    label: "Turn video off",
    kind: "secondary",
    Icon: icons.VideoCallSolidIcon,
  },
};

export const VideoOff = {
  args: {
    label: "Turn video on",
    kind: "primary",
    Icon: icons.VideoCallOffSolidIcon,
  },
};

export const MicOn = {
  args: {
    label: "Turn mic off",
    kind: "secondary",
    Icon: icons.MicOnSolidIcon,
  },
};

export const MicOff = {
  args: {
    label: "Turn mic on",
    kind: "primary",
    Icon: icons.MicOffSolidIcon,
  },
};

export const ShareScreen = {
  args: {
    label: "Share screen",
    kind: "secondary",
    Icon: icons.ShareScreenSolidIcon,
  },
};

export const SharingScreen = {
  args: {
    label: "Stop sharing screen",
    kind: "primary",
    Icon: icons.ShareScreenSolidIcon,
  },
};

export const Fullscreen = {
  args: {
    label: "Full screen",
    kind: "secondary",
    Icon: icons.ExpandIcon,
  },
};

export const EndCall = {
  args: {
    label: "End call",
    kind: "primary",
    destructive: true,
    Icon: icons.EndCallIcon,
  },
};

export const Invite = {
  args: {
    label: "Invite",
    kind: "primary",
    size: "sm",
    Icon: icons.UserAddIcon,
  },
};

export const Edit = {
  args: {
    label: "Edit",
    kind: "primary",
    size: "sm",
    Icon: icons.EditIcon,
  },
};

export const Search = {
  args: {
    label: "Search",
    kind: "secondary",
    size: "sm",
    Icon: icons.SearchIcon,
  },
};

export const Save = {
  args: {
    label: "Save",
    kind: "primary",
    size: "sm",
    Icon: icons.CheckIcon,
  },
};

export const SaveDisabled = {
  args: {
    label: "Save",
    kind: "primary",
    size: "sm",
    Icon: icons.CheckIcon,
    disabled: true,
  },
};

export const Cancel = {
  args: {
    label: "Cancel",
    kind: "secondary",
    size: "sm",
    Icon: icons.CloseIcon,
  },
};

export const Close = {
  args: {
    label: "Close",
    kind: "secondary",
    size: "sm",
    Icon: icons.CloseIcon,
  },
};
