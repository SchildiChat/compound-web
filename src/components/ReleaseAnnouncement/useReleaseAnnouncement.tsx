/*
Copyright 2024 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import {
  arrow,
  autoUpdate,
  limitShift,
  offset,
  Placement,
  shift,
  useFloating,
  useId,
  useInteractions,
  useRole,
} from "@floating-ui/react";
import { MouseEventHandler, useMemo, useRef } from "react";

interface UseReleaseAnnouncementProps {
  /**
   * Open or not the release announcement.
   */
  open: boolean;
  /**
   * The header of the release announcement.
   */
  header: string;
  /**
   * The description of the release announcement.
   */
  description: string;
  /**
   * The label of the close button.
   */
  closeLabel: string;
  /**
   * The placement of the release announcement.
   */
  placement: Placement;
  /**
   * The event handler for the close button.
   */
  onClick: MouseEventHandler<HTMLButtonElement>;
}

/**
 * This hook provides the necessary props to create a release announcement.
 */
export function useReleaseAnnouncement({
  open,
  header,
  description,
  closeLabel,
  placement,
  onClick,
}: UseReleaseAnnouncementProps) {
  // Set on `aria-labelledby` attribute
  const labelId = useId();
  // Set on `aria-describedby` attribute
  const descriptionId = useId();
  const arrowRef = useRef(null);

  const data = useFloating({
    placement,
    open,
    whileElementsMounted: autoUpdate,
    middleware: [
      // arrow height 12px + 4px padding
      offset(16),
      shift({ limiter: limitShift({ offset: 50 }) }),
      // add the little arrow along with the floating content
      arrow({
        element: arrowRef,
      }),
    ],
  });

  // The role is used to determine the correct aria role for the floating content.
  const role = useRole(data.context);
  const interactions = useInteractions([role]);

  // Put in the context all the useful data and interactions
  return useMemo(
    () => ({
      open,
      ...data,
      ...interactions,
      labelId,
      descriptionId,
      header,
      description,
      closeLabel,
      onClick,
      arrowRef,
    }),
    [
      open,
      role,
      interactions,
      data,
      labelId,
      descriptionId,
      header,
      description,
      closeLabel,
      onClick,
      arrowRef,
    ],
  );
}
