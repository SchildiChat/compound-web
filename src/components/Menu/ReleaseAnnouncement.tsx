/*
Copyright 2024 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, {
  cloneElement,
  createContext,
  Dispatch,
  isValidElement,
  PropsWithChildren,
  SetStateAction,
  useContext,
  useRef,
  useState,
} from "react";
import {
  autoUpdate,
  flip,
  FloatingPortal,
  FloatingFocusManager,
  offset,
  Placement,
  shift,
  useFloating,
  useInteractions,
  useMergeRefs,
  useRole,
  useId,
  FloatingArrow,
  arrow,
} from "@floating-ui/react";
import { Text } from "../Typography/Text";
import { Button } from "../Button";
import styles from "./ReleaseAnnouncement.module.css";

interface ReleaseAnnouncementProps {
  header: string;
  description: string;
  closeLabel: string;
  placement?: Placement;
}

export function ReleaseAnnouncement({
  children,
  placement = "bottom",
  ...props
}: PropsWithChildren<ReleaseAnnouncementProps>) {
  const context = useReleaseAnnouncement({ placement, ...props });

  return (
    <ReleaseAnnouncementContext.Provider value={context}>
      <ReleaseAnnouncementAnchor>{children}</ReleaseAnnouncementAnchor>
      <ReleaseAnnouncementContainer>
        <ReleaseAnnouncementContent />
      </ReleaseAnnouncementContainer>
    </ReleaseAnnouncementContext.Provider>
  );
}

interface UseReleaseAnnouncementProps {
  header: string;
  description: string;
  closeLabel: string;
  placement: Placement;
  onClose?: () => void;
}

function useReleaseAnnouncement({
  header,
  description,
  closeLabel,
  placement,
  onClose,
}: UseReleaseAnnouncementProps) {
  const [open, setOpen] = useState(true);
  const [labelId, setLabelId] = useState<string | undefined>();
  const [descriptionId, setDescriptionId] = useState<string | undefined>();
  const arrowRef = useRef(null);

  const data = useFloating({
    placement,
    open,
    onOpenChange: setOpen,
    whileElementsMounted: autoUpdate,
    middleware: [
      // arrow height 12px + 4px padding
      offset(16),
      flip({
        crossAxis: placement.includes("-"),
        fallbackAxisSideDirection: "end",
        padding: 5,
      }),
      shift(),
      arrow({
        element: arrowRef,
      }),
    ],
  });

  const context = data.context;

  const role = useRole(context);
  const interactions = useInteractions([role]);

  return React.useMemo(
    () => ({
      open,
      setOpen,
      ...interactions,
      ...data,
      labelId,
      descriptionId,
      setLabelId,
      setDescriptionId,
      header,
      description,
      closeLabel,
      onClose,
      arrowRef,
    }),
    [
      open,
      setOpen,
      interactions,
      data,
      labelId,
      descriptionId,
      header,
      description,
      closeLabel,
      onClose,
      arrowRef,
    ],
  );
}

type ContextType =
  | (ReturnType<typeof useReleaseAnnouncement> & {
      setLabelId: Dispatch<SetStateAction<string | undefined>>;
      setDescriptionId: Dispatch<SetStateAction<string | undefined>>;
    })
  | null;

const ReleaseAnnouncementContext = createContext<ContextType>(null);

const useReleaseAnnouncementContext = () => {
  const context = useContext(ReleaseAnnouncementContext);

  if (context == null) {
    throw new Error(
      "ReleaseAnnouncement components must be wrapped in <ReleaseAnnouncement />",
    );
  }

  return context;
};

interface ReleaseAnnouncementAnchorProps {}

function ReleaseAnnouncementAnchor({
  children,
}: PropsWithChildren<ReleaseAnnouncementAnchorProps>) {
  const context = useReleaseAnnouncementContext();
  const childrenRef = (children as any).ref;
  const ref = useMergeRefs([context.refs.setReference, childrenRef]);

  if (!isValidElement(children)) {
    throw new Error(
      "ReleaseAnnouncement anchor must be a single valid React element",
    );
  }

  return cloneElement(
    children,
    context.getReferenceProps({
      ref,
      ...children.props,
      "data-state": context.open ? "open" : "closed",
    }),
  );
}

interface ReleaseAnnouncementContainerProps {}

function ReleaseAnnouncementContainer({
  children,
}: PropsWithChildren<ReleaseAnnouncementContainerProps>) {
  const {
    context: floatingContext,
    arrowRef,
    ...context
  } = useReleaseAnnouncementContext();

  if (!floatingContext.open) return null;

  return (
    <FloatingPortal>
      <FloatingFocusManager context={floatingContext}>
        <div
          ref={context.refs.setFloating}
          style={context.floatingStyles}
          aria-labelledby={context.labelId}
          aria-describedby={context.descriptionId}
          {...context.getFloatingProps()}
          className={styles.content}
        >
          <FloatingArrow
            ref={arrowRef}
            context={floatingContext}
            width={20}
            height={12}
          />
          {children}
        </div>
      </FloatingFocusManager>
    </FloatingPortal>
  );
}

function ReleaseAnnouncementContent() {
  const {
    setLabelId,
    setDescriptionId,
    header,
    description,
    closeLabel,
    onClose,
    setOpen,
  } = useReleaseAnnouncementContext();
  const labelId = useId();
  const descriptionId = useId();

  // Only sets `aria-describedby` and `aria-labelledby` on the ReleaseAnnouncementContainer component
  // if this component is mounted inside it.
  React.useLayoutEffect(() => {
    setLabelId(labelId);
    setDescriptionId(descriptionId);
    return () => {
      setLabelId(undefined);
      setDescriptionId(undefined);
    };
  }, [labelId, descriptionId, setLabelId, setDescriptionId]);

  return (
    <>
      <Text
        as="h3"
        id={labelId}
        className={styles.heading}
        size="lg"
        weight="semibold"
      >
        {header}
      </Text>
      <Text
        as="span"
        id={descriptionId}
        className={styles.description}
        size="sm"
        weight="regular"
      >
        {description}
      </Text>
      <Button
        size="sm"
        kind="secondary"
        className={styles.button}
        onClick={() => {
          setOpen(false);
          onClose?.();
        }}
      >
        {closeLabel}
      </Button>
    </>
  );
}
