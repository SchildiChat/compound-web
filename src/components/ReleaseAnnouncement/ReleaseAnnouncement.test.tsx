/*
 *
 * Copyright 2024 New Vector Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * /
 */

import { composeStories } from "@storybook/react";
import * as stories from "./ReleaseAnnouncement.stories";
import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import React from "react";
import { ReleaseAnnouncement } from "./ReleaseAnnouncement";
import { Button } from "../Button";

const { Default, MultiLinesContent, BottomPlacement } = composeStories(stories);

describe("ReleaseAnnouncement", () => {
  it("renders", async () => {
    render(<Default />);
    expect(await screen.findByRole("dialog")).toMatchSnapshot();
  });

  it("renders with a long label header and description", async () => {
    render(<MultiLinesContent />);
    expect(await screen.findByRole("dialog")).toMatchSnapshot();
  });

  it("renders the component at the bottom of the trigger button", async () => {
    render(<BottomPlacement />);
    expect(await screen.findByRole("dialog")).toMatchSnapshot();
  });

  it("renders the component when closed", async () => {
    render(
      <ReleaseAnnouncement
        open={false}
        header="header"
        description="description"
        closeLabel="Ok"
        onClick={vi.fn()}
      >
        <Button>Open</Button>
      </ReleaseAnnouncement>,
    );
    expect(screen.queryByRole("dialog")).toBeNull();
  });
});
