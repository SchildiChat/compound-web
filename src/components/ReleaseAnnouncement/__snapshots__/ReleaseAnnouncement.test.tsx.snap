// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ReleaseAnnouncement > renders 1`] = `
<div
  data-floating-ui-inert=""
>
  <button
    aria-controls=":r2:"
    aria-expanded="true"
    aria-haspopup="dialog"
    aria-label="open menu"
    class="_button_2a1efe"
    data-kind="primary"
    data-size="lg"
    data-state="open"
    role="button"
    tabindex="0"
  >
    Open menu
  </button>
  <span
    data-floating-ui-focus-guard=""
    data-type="outside"
    role="button"
    style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: fixed; white-space: nowrap; width: 1px; top: 0px; left: 0px;"
    tabindex="0"
  />
  <span
    aria-owns=":r4:"
    style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: fixed; white-space: nowrap; width: 1px; top: 0px; left: 0px;"
  />
  <span
    data-floating-ui-focus-guard=""
    data-type="outside"
    role="button"
    style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: fixed; white-space: nowrap; width: 1px; top: 0px; left: 0px;"
    tabindex="0"
  />
</div>
`;

exports[`ReleaseAnnouncement > renders the component at the bottom of the trigger button 1`] = `
<div
  data-floating-ui-inert=""
>
  <button
    aria-controls=":re:"
    aria-expanded="true"
    aria-haspopup="dialog"
    aria-label="open menu"
    class="_button_2a1efe"
    data-kind="primary"
    data-size="lg"
    data-state="open"
    role="button"
    tabindex="0"
  >
    Open menu
  </button>
  <span
    data-floating-ui-focus-guard=""
    data-type="outside"
    role="button"
    style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: fixed; white-space: nowrap; width: 1px; top: 0px; left: 0px;"
    tabindex="0"
  />
  <span
    aria-owns=":rg:"
    style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: fixed; white-space: nowrap; width: 1px; top: 0px; left: 0px;"
  />
  <span
    data-floating-ui-focus-guard=""
    data-type="outside"
    role="button"
    style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: fixed; white-space: nowrap; width: 1px; top: 0px; left: 0px;"
    tabindex="0"
  />
</div>
`;

exports[`ReleaseAnnouncement > renders with a long label header and description 1`] = `
<div
  data-floating-ui-inert=""
>
  <button
    aria-controls=":r8:"
    aria-expanded="true"
    aria-haspopup="dialog"
    aria-label="open menu"
    class="_button_2a1efe"
    data-kind="primary"
    data-size="lg"
    data-state="open"
    role="button"
    tabindex="0"
  >
    Open menu
  </button>
  <span
    data-floating-ui-focus-guard=""
    data-type="outside"
    role="button"
    style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: fixed; white-space: nowrap; width: 1px; top: 0px; left: 0px;"
    tabindex="0"
  />
  <span
    aria-owns=":ra:"
    style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: fixed; white-space: nowrap; width: 1px; top: 0px; left: 0px;"
  />
  <span
    data-floating-ui-focus-guard=""
    data-type="outside"
    role="button"
    style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: fixed; white-space: nowrap; width: 1px; top: 0px; left: 0px;"
    tabindex="0"
  />
</div>
`;
